File: tools/scripts/nrql-query.py
#!/usr/bin/env python3
"""nrql-query tool — runs a NRQL query via New Relic NerdGraph.

Reads the tool input as JSON on stdin (and args as UPPERCASE env vars), uses the
NEW_RELIC_USER_API_KEY supplied AT POINT OF USE by the platform broker (never stored
in the repo), and prints the rows as JSON on stdout. Non-zero exit on failure.
"""
import json, os, sys, urllib.request

def main() -> int:
    raw = sys.stdin.read() or "{}"
    args = json.loads(raw)
    nrql = args.get("nrql") or os.environ.get("NRQL")
    if not nrql:
        print("missing 'nrql'", file=sys.stderr); return 2
    key = os.environ.get("NEW_RELIC_USER_API_KEY")
    acct = os.environ.get("NEW_RELIC_ACCOUNT_ID")
    if not key or not acct:
        print("NEW_RELIC_USER_API_KEY / NEW_RELIC_ACCOUNT_ID not provided", file=sys.stderr); return 3
    query = (
        '{ actor { account(id: %s) { nrql(query: %s) { results } } } }'
        % (acct, json.dumps(nrql))
    )
    req = urllib.request.Request(
        "https://api.newrelic.com/graphql",
        data=json.dumps({"query": query}).encode(),
        headers={"Content-Type": "application/json", "API-Key": key},
    )
    try:
        with urllib.request.urlopen(req, timeout=int(args.get("timeout_seconds", 20))) as r:
            body = json.loads(r.read())
    except Exception as e:  # ponytail: surface the error as tool failure
        print(f"NRQL request failed: {e}", file=sys.stderr); return 1
    rows = (((body.get("data") or {}).get("actor") or {}).get("account") or {}).get("nrql", {}).get("results", [])
    print(json.dumps({"rows": rows, "total": len(rows)}))
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: tools/scripts/nrql-query.mjs
#!/usr/bin/env node
// nrql-query tool — runs a NRQL query via New Relic NerdGraph.
// Reads tool input as JSON on stdin, uses NEW_RELIC_USER_API_KEY /
// NEW_RELIC_ACCOUNT_ID from env (injected by the platform), prints rows as JSON.
import https from "node:https";

const raw = await new Promise((r) => {
  let s = "";
  process.stdin.on("data", (c) => (s += c));
  process.stdin.on("end", () => r(s || "{}"));
});
const args = JSON.parse(raw);
const nrql = args.nrql || process.env.NRQL;
if (!nrql) { console.error("missing 'nrql'"); process.exit(2); }
const key = process.env.NEW_RELIC_USER_API_KEY, acct = process.env.NEW_RELIC_ACCOUNT_ID;
if (!key || !acct) { console.error("NEW_RELIC_USER_API_KEY / NEW_RELIC_ACCOUNT_ID not provided"); process.exit(3); }

const query = `{ actor { account(id: ${acct}) { nrql(query: ${JSON.stringify(nrql)}) { results } } } }`;
const body = JSON.stringify({ query });
const req = https.request("https://api.newrelic.com/graphql", {
  method: "POST",
  headers: { "Content-Type": "application/json", "API-Key": key, "Content-Length": Buffer.byteLength(body) },
  timeout: (args.timeout_seconds || 20) * 1000,
}, (res) => {
  let d = "";
  res.on("data", (c) => (d += c));
  res.on("end", () => {
    try {
      const rows = JSON.parse(d)?.data?.actor?.account?.nrql?.results ?? [];
      console.log(JSON.stringify({ rows, total: rows.length }));
    } catch (e) { console.error("NRQL parse failed: " + e.message); process.exit(1); }
  });
});
req.on("error", (e) => { console.error("NRQL request failed: " + e.message); process.exit(1); });
req.on("timeout", () => { req.destroy(); console.error("NRQL request timed out"); process.exit(1); });
req.end(body);
